_core/smart/clarityRef.md

```diff
@@ -10,104 +10,126 @@ This file contains the reference for the Clarity language.
+for more details on Stacks transactions), public function may be called
+by other smart contracts.
+
+Public functions _must_ return a `(response ...)` type. This is used
+by Clarity to determine whether or not to materialize any changes from
+the execution of the function. If a function returns an `(err ...)`
+type, and mutations on the blockchain state from executing the
+function (and any function that it called during execution) will be
+aborted.
+
+In addition to function defined via `define-public`, contracts may expose
+read-only functions. These functions, defined via `define-read-only`, are
+callable by other smart contracts, and may be queryable via public blockchain
+explorers. These functions _may not_ mutate any blockchain state. Unlike normal
+public functions, read-only functions may return any type.
+
+## Contract Calls
+
+A smart contract may call functions from other smart contracts using a
+`(contract-call?)` function.
+
+This function returns a response type result-- the return value of the
+called smart contract function.
+
+We distinguish 2 different types of `contract-call?`:
+
+* Static dispatch: the callee is a known, invariant contract available
+on-chain when the caller contract is deployed. In this case, the
+callee's principal is provided as the first argument, followed by the
+name of the method and its arguments:
+
+```scheme
+(contract-call?
+    .registrar
+    register-name
+    name-to-register)
 ```
 
-This allows for creating named tuples on the fly, which is useful for data maps where the keys and values are themselves named tuples. Values in a given mapping are set or fetched using:
-
-<table class="uk-table uk-table-small">
-<tr>
-  <th class="uk-width-small">Function</th>
-  <th>Description</th>
-</tr>
-<tr>
-  <td><code>(map-get map-name key-tuple)</code></td>
-  <td>Fetches the value associated with a given key in the map, or returns <code>none</code> if there is no such value.</td>
-</tr>
-<tr>
-   <td><code>(map-set! map-name key-tuple value-tuple)</code></td>
-  <td>Sets the value of key-tuple in the data map</td>
-</tr>
-  <tr>
-   <td><code>(map-insert! map-name key-tuple value-tuple)</code></td>
-  <td>Sets the value of key-tuple in the data map if and only if an entry does not already exist.</td>
-</tr>
-  <tr>
-   <td><code>(map-delete! map-name key-tuple)</code></td>
-  <td>Deletes key-tuple from the data map.</td>
-</tr>
-</table>
-
-
-To access a named value of a given tuple, the `(get name tuple)` function returns that item from the tuple.
-
-### Optional type
-
-Represents an optional value. This is used in place of the typical usage of "null" values in other languages, and represents a type that can either be some value or `none`. Optional types are used as the return types of data-map functions.
+* Dynamic dispatch: the callee is passed as an argument, and typed
+as a trait reference (<A>).
+
+```scheme
+(define-public (swap (token-a <can-transfer-tokens>)
```

Comment:
One thing I've been curious about - if I was to actually call this function, would the first argument be a fully qualified contract principal?

Reply:
Can I also do dynamic dispatch without a trait reference? I.e. `token-a` is just a `principal`. I can still do a `contract-call?` in the function, yeah?

Reply:
> One thing I've been curious about - if I was to actually call this function, would the first argument be a fully qualified contract principal?

Yes -- it is either a _literal_ contract principal or a matching trait variable.

> Can I also do dynamic dispatch without a trait reference? I.e. token-a is just a principal. I can still do a contract-call? in the function, yeah?

No -- that's what differentiates the trait execution from general dynamic dispatch. `contract-call` can only be invoked either on a contract-literal or a trait parameter. This ensures that the type system can enforce the invocation during analysis, rather than run time.